content/2015.08.notifications-preamble.md

```diff
@@ -0,0 +1,174 @@
+            Consumer [shape=box];
+            endpoint -> Consumer [label=4, dir=both];
+        }
+    }
+%}
+Pros:
+
+* Kinto could be shipped as a "turn-key" solution with a fully functioning stack baked right in; while there would be a queuing service involved ([Celery](http://www.celeryproject.org/), for example), it would be completely abstracted by Kinto.
+* Since the queuing mechanism is integral, Kinto would have full visibility and control over processing and delivery.
+* Apart from the queuing service there would be no additional services or software to deploy.
+
+Cons:
+
+* This would involve a fairly serious increase in the scope, complexity, and size of the Kinto code base.
+* The tight coupling means that Kinto, in its entirety, would need to be re-deployed if changes were necessary in the queuing service.
+* Similarly, scaling the queue would imply scaling Kinto itself
+
+This model has certain interesting advantages; however, the relative volume of additional code that would need to be developed is reason enough to give us pause ([scope creep](https://en.wikipedia.org/wiki/Scope_creep) is a real problem). Like the previous model, this also runs afoul of our tech agnosticism guideline since we would be locking Kinto to a given queuing service directly. Finally, we have (admittedly theoretical) concerns about how this model would scale - it feels like a waste to increase two resources when only one needs to grow, for example.
+
+## The Broker
+
+The final approach is a bit of a mix of the previous two. In this model, we implement a formal "broker" that is part of the Kinto *eco-system*, but external to Kinto itself. The broker could be anything from a few lines of Python sitting on a [pipe](https://en.wikipedia.org/wiki/Named_pipe) to a proper micro-service stack. This decouples the queuing mechanism from Kinto while preserving a strong *promise* of interoperability and control.
+
+{% graphviz
+    dot {
+        digraph broker {
+            rankdir=LR;
+            Kinto [shape=box];
+            subgraph cluster_broker {
+                style=filled;
+                color=lightgrey;
+                node [style=filled, color=white];
+                Broker [shape=box];
+                worker -> Broker [label="2. subscribe"];
+                Broker -> worker [label="3. async"];
+            }
+            Kinto -> Broker [label="1. POST; sync"];
+            Broker -> Kinto [label="feedback loop", style=dotted, dir=both];
+            Consumer [shape=box];
+            worker -> Consumer [label=4, dir=both];
+        }
+    }
+%}
+
+Pros:
+
+* Provides a clean and easily understandable separation of responsibilities.
+* Frees both Kinto and the consumer from any particular service dependency.
+* Permits the broker to scale independently from Kinto.
+* Requires only a moderate amount of new code.
+* Effectively allows for a highly flexible solution where just about anything imaginable could be plugged in.
```

Comment:
I would also add that the decoupling the broker model simplifies a lot error management when its specific to the use case. e.g. in Balrog use case, dealing with problems like "I nee to try sending again and again the event to Balrog until it works" or "if I get more than 10 write events within an hour, let's merge them into a single event for Balrog"